/**
 Copyright 2022 PrimeQA Team

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

syntax = "proto3";
import "parameter.proto";
import "indexer.proto";

package retrieve;

/**
    Service to process and retrieve documents
*/
service Retriever {
    rpc GetRetrievers(GetRetrieversRequest) returns (GetRetrieversResponse);
    rpc Retrieve(RetrieveRequest) returns (RetrieveResponse);
}

message RetrieverComponent {
    string retriever_id = 1;
    repeated parameter.Parameter parameters = 2;
}

message GetRetrieversRequest {
}

message GetRetrieversResponse {
    repeated RetrieverComponent retrievers = 1;
}


message RetrieveRequest {
    RetrieverComponent retriever = 1;
    string index_id = 2; 
    repeated string queries = 3;
}

message Hit {
    index.Document document = 1;
    double score = 2;
}

message HitPerQuery {
    repeated Hit hits = 1;
}

message RetrieveResponse {
    repeated HitPerQuery hits = 1;
}
